src/audio/module_adapter/module_adapter.c

```diff
@@ -165,6 +165,55 @@ struct comp_dev *module_adapter_new(const struct comp_driver *drv,
 	return NULL;
 }
 
+static int module_adapter_sink_src_prepare(struct comp_dev *dev)
+{
+	struct comp_buffer __sparse_cache *source_buffers_c[PLATFORM_MAX_STREAMS];
+	struct comp_buffer __sparse_cache *sinks_buffers_c[PLATFORM_MAX_STREAMS];
+	struct sink __sparse_cache *audio_sink[PLATFORM_MAX_STREAMS];
+	struct source __sparse_cache *audio_src[PLATFORM_MAX_STREAMS];
+	struct processing_module *mod = comp_get_drvdata(dev);
+	struct list_item *blist;
+	uint32_t num_of_sources = 0;
+	uint32_t num_of_sinks = 0;
+	int ret;
+
+	/* acquire all sink and source buffers, get handlers to sink/source API */
+	list_for_item(blist, &dev->bsink_list) {
+		struct comp_buffer *sink_buffer_uc;
+
+		sink_buffer_uc = container_of(blist, struct comp_buffer, source_list);
+		sinks_buffers_c[num_of_sinks] = buffer_acquire(sink_buffer_uc);
+		audio_sink[num_of_sinks] =
+				audio_stream_get_sink(&sinks_buffers_c[num_of_sinks]->stream);
+		sink_reset_num_of_processed_bytes(audio_sink[num_of_sinks]);
+		num_of_sinks++;
+	}
+
+	list_for_item(blist, &dev->bsource_list) {
+		struct comp_buffer *source_buffer_uc;
+
+		source_buffer_uc = container_of(blist, struct comp_buffer, sink_list);
+		source_buffers_c[num_of_sources] = buffer_acquire(source_buffer_uc);
+		audio_src[num_of_sources] =
+				audio_stream_get_source(&source_buffers_c[num_of_sources]->stream);
+		source_reset_num_of_processed_bytes(audio_src[num_of_sources]);
+		num_of_sources++;
+	}
+
+	/* Prepare module */
+	ret = module_prepare(mod, audio_src, num_of_sources, audio_sink, num_of_sinks);
+
+	/* release all source buffers in reverse order */
```

Comment:
why reverse order?

Reply:
According to @lyakh  it has to be this way. I don't fully understand why,

Reply:
@kv2019i In general you don't do
```
lock(a);
lock(b);
...
unlock(a);
unlock(b);
```
but in the Zephyr spinlock case the reason is also rather simple: `k_spin_unlock()` restores the CPU state to the one saved by the respective `k_spin_lock()`. So typically you take the first spinlock with interrupts enabled, the second one, however, already with interrupts disabled. So if you then unlock in the wrong order, you first restore to the state with interrupts enabled, which is too early at that point, but then again to interrupts disabled, and you leave your context that way - with interrupts still disabled.